File: cpp/ipc/pipe/pipe.cpp
#include <fcntl.h>
#include <unistd.h>
#include <cstdio>

#include <string>
#include <memory>
#include <chrono>

#include "../utils.hpp"

int main(int argc, char* argv[])
{
    // Pipe fd: fd[0] - read, fd[1] - write
    int fd[2];
    int feedbackfd[2];

    std::unique_ptr<struct Data> data(new struct Data);
    size_t counter = 0;

    if (pipe(fd) == -1)
    {
        perror("pipe error");
        return EXIT_FAILURE;
    }

    if (pipe(feedbackfd) == -1)
    {
        perror("feedback error");
        return EXIT_FAILURE;
    }

    pid_t pid = fork();
    // Base
    if (pid > 0)
    {
        close(fd[0]); // Close read
        close(feedbackfd[1]);  // Close feedback write 
        
        data->iSignal = 0;
        strcpy(data->buffer, message.c_str());
        for (size_t i = 0; i < (BUFFER_LEN / message.size() - 1); ++i)
        {
            strcat(data->buffer, message.c_str());
        }

        while (counter < RUN_TIMES)
        {
            if (data->iSignal != 1)
            {
                data->iSignal = 1;
                
                write(fd[1], data.get(), sizeof(data->iSignal) + BUFFER_LEN);
#ifdef DEBUG
                printf("[%ld] write text (%ld bytes) to pipe: %s\n", counter, strlen(data->buffer), data->buffer);
#endif
                data->iSignal = 0;

                ++counter;
            }
            read(feedbackfd[0], data.get(), sizeof(data->iSignal));
        }

        close(fd[1]);
        close(feedbackfd[0]);
    }
    // Derived
    else if (pid == 0)
    {
        close(fd[1]);  // Close fd write
        close(feedbackfd[0]);  // Close feedback read 
 
        size_t size = 0;
        data->iSignal = 1;
        auto start = std::chrono::high_resolution_clock::now();
        while (counter < RUN_TIMES)
        {
            read(fd[0], data.get(), sizeof(data->iSignal) + BUFFER_LEN);
            if (data->iSignal != 0)
            {
                size = strlen(data->buffer);
#ifdef DEBUG
                printf("[%ld] Current txt (%ld bytes): %s\n", counter, size, data->buffer);
#endif
                data->iSignal = 0;
                write(feedbackfd[1], data.get(), sizeof(data->iSignal));

                ++counter;
            }
        }
        auto end = std::chrono::high_resolution_clock::now();
        auto duration = std::chrono::duration<double>(end - start);
        printf("Process %ld bytes package %ld times, cost %fs\n", size, counter, duration.count());

        close(fd[0]);
        close(feedbackfd[1]);
    }

    return EXIT_SUCCESS;
}

File: python3/processes/ipc/pipe/pipe.py
import time
from multiprocessing import Process
from multiprocessing import Pipe


RUN_TIMES = 1000000
BUFFER_LEN = 10240

# 64bytes
MESSAGE = "QWERTY123456789qwertyuiopasdfghjklzxcvbnm,./';\\][=-+_)(*&^%$#@!~"

def producer(dataConnection, ctrlConnection):
    iSignal = 0
    data = MESSAGE * int(BUFFER_LEN / len(MESSAGE) - 1)

    counter = 0
    while counter < RUN_TIMES:
        if iSignal != 1:
            iSignal = 1
            dataConnection.send(data)
            # print("[{0}] write text ({1} bytes) to pipe".format(counter, len(data)))
            # print("text: {0}".format(data))

            counter += 1
        iSignal = ctrlConnection.recv()

def consumer(dataConnection, ctrlConnection):
    counter = 0
    doneSignal = 0
    start = time.time()
    while counter < RUN_TIMES:
        data = dataConnection.recv()
        if len(data) > 0:
            # print("[{0}] Current text ({1} bytes)".format(counter, len(data)))
            ctrlConnection.send(doneSignal)
            data = ""
            counter += 1
    end = time.time()
    print("Process {0} times in {1} sec".format(counter, end - start))


dataConn1, dataConn2 = Pipe(duplex=True)  # 1 for read, 2 for write
ctrlConn1, ctrlConn2 = Pipe(duplex=True)
sender = Process(target=producer, args=(dataConn2, ctrlConn1))
recver = Process(target=consumer, args=(dataConn1, ctrlConn2))

sender.start()
recver.start()

sender.join()
recver.join()

